system/system.proto

```diff
@@ -227,6 +232,86 @@ message PingResponse {
 }
 
+// A ProbeRequest describes the probe operation to perform.  The
+// destination field and exactly one of the probe_target fields
+// are required.  Any field not specified is set to a reasonable server
+// specified value.  Not all fields are supported by all vendors.
+message ProbeRequest {
+  string destination = 1;   // Proxy address to probe. required.
+  string source = 2;        // Source address to probe from.
+  oneof probe_target {
+    // exactly one of the next 4 items must be present.
+    int32 ifindex = 3;        // ifIndex to probe, if present.
+    string ifname = 4;        // ifName to probe, if present.
+    string ifaddr = 5;        // ifAddr to probe, if present.
+    string remote = 6;        // remote address to probe, if present.
+  }
+  int32 count = 7;          // Number of requests to make, default 3.
+  int64 interval = 8;       // Nanoseconds between requests.
+  int64 wait = 9;           // Nanoseconds to wait for a response.
+
+  string network_instance = 10; // Network instance to probe the proxy node in.
+}
+
+// A ProbeResponse represents either the response to a single probe packet
+// (the code field is non-zero) or the summary statistics (sent is non-zero).
+//
+// For a single probe packet, time is the round trip time, in nanoseconds.  For
+// summary statistics, it is the time spent by the ping operation.  The time is
+// not always present in summary statistics.
+//
+// In the summary, each map is keyed by the enum values referred to by that map.
+// The codes map is keyed by the Code enum, the states map is keyed by the State
+// enum, and the actives map is keyed by the Active enum.
+message ProbeResponse {
+  enum Code {
+    CODE_UNKNOWN = 0;
+    CODE_NO_ERROR = 1;
+    CODE_MALFORMED_QUERY = 2;
+    CODE_NO_SUCH_INTERFACE = 3;
+    CODE_NO_SUCH_TABLE_ENTRY = 4;
+    CODE_MULTIPLE_INTERFACES = 5;
+  }
+  enum State {
+    STATE_UNKNOWN = 0;
+    STATE_INCOMPLETE = 1;
+    STATE_REACHABLE = 2;
+    STATE_STALE = 3;
+    STATE_DELAY = 4;
+    STATE_PROBE = 5;
+    STATE_FAILED = 6;
+  }
+  enum Active {
+    ACTIVE_UNKNOWN = 0;
+    ACTIVE_INACTIVE = 1;
+    ACTIVE_NOV4_NOV6 = 2;
+    ACTIVE_V4 = 3;
+    ACTIVE_V6 = 4;
+    ACTIVE_V4_V6 = 5;
+  }
+  string source = 1;
+  int64 time = 2;
+
+  // For individual responses
+  int32 sequence = 3;
+  Code code = 4;
+  // state is only present if code is CODE_NO_ERROR
+  // and request was remote
+  State state = 5;
+  // active is only present if code is CODE_NO_ERROR
+  // and request was not remote
+  Active active = 6;
+
+  // summary
+  int32 sent = 7;       // Total packets sent.
+  int32 errors = 8;     // Count of errors received when sending packets.
+  int32 received = 9;   // Total packets received.
+  // histograms in summary
+  map <uint32,uint32> codes = 10;     // Keys are values of Code enum
+  map <uint32,uint32> states = 11;    // Keys are values of State enum
+  map <uint32,uint32> actives = 12;   // Keys are values of Active enum
```

Comment:
Is this really needed?
gNOI client can do the math, adding a summary to each response seems unnecessary

Reply:
That's true - other than the errors, all of these values can be derived from the set of responses. I copied this concept from the "ping" response, which also includes summary values that can be derived from the set of responses. If it's useful there, then I thought it would be useful here.

Reply:
Personally, I think the direct streaming of responses is a better approach, but if you want to provide an option to return the summary, I would suggest refactoring this to use the `oneof` model and splitting it into two message types.

Reply:
As I said, I modeled this after "ping" to keep it similar since the protocols are similar. Should "ping" also use `oneof`?

Reply:
>Should "ping" also use oneof?

Maybe :-) But it's a bit too late to change that, and not the point. 

In this case, the state machine is significantly more complex than in ping rpc. IMO it's easier to write a client that doesn't have to infer the meaning of the message (a summary vs an individual response) based on field values, especially considering that you mostly have independent fields in each case. That also automatically answers questions such as "what fields are expected or allowed to be set together, what are valid/invalid combinations". 

Is there any concern about this?

Reply:
I can make the change, I'm just concerned about creating a difference from the ping RPC.  I figured that the ping definition established how this kind of RPC was desired to be expressed in GNOI.  (For example, I used an int64 for the time, even though uint64 is a better fit for an absolute timestamp, because that's what ping did).